### DISCUSSION TITLE
Support for "Online" LLM

### DISCUSSION THREAD

**Discussion by cpursley** [brainlid/langchain#100] [Ideas]
Maybe out of scope but you know what would be cool - doing this sort of thing (Perplexity clone ie., "online" live LLM):

- https://news.ycombinator.com/item?id=39923404
- https://github.com/nilsherzig/LLocalSearch/discussions/38 (architecture)
 -https://github.com/searxng/searxng (open source search aggregator)

Basically a RAG approach that knows when to run a live search through the searxng aggregator. Perhaps combined with pgvector and the bumblebee harness GPU for a true open source setup. And maybe hooked up to [Crawly](https://github.com/brainlid/langchain/pull/22) to get the page content in parallel.

Could make for an interesting open source fly setup (as all the services could be run there).

Or a simpler version using the Bing API: https://github.com/brainlid/langchain/issues/48

**Comment by brainlid** [comment]
I think it's a cool idea.

**Comment by cpursley** [comment]
Just came across this interesting project:

- https://docs.tavily.com/docs/tavily-api/introduction
- https://github.com/assafelovic/gpt-researcher

**Comment by cpursley** [comment]
I'm thinking through approaches for building "**websearch**". Perhaps as another tool (like the calculator)? Where there's a swapable search engine adapter ([Bing API](https://www.microsoft.com/en-us/bing/apis/bing-web-search-api), [Brave](https://brave.com/search/api/), [searxng](https://github.com/searxng/searxng), etc).

And where you can specify via confg how many results it should fetch in parallel. But it's unclear how to best handle the next step - [semantically chunk](https://github.com/revelrylabs/text_chunker_ex) the html and vectorize it in memory via [Nx](https://github.com/elixir-nx/hnswlib)? Or dump into Postgres pgvector?

Then there's some other things that might need to be done for better reliability like this video describes (regarding the Grader and Hallucination checker): [Reliable, fully local RAG agents with LLaMA3](https://www.youtube.com/watch?v=-ROS6gfYIts)

![websearch](https://github.com/brainlid/langchain/assets/1223773/9be729ed-965a-4a6c-866f-1b9424f069ca)

There's also some interesting ideas here: https://docs.tavily.com/blog/building-openai-assistant

Thoughts?

**Reply by cpursley** [reply]
I'm also perplexed on how to go about having the llm understand when it should search the web or not.

**Reply by brainlid** [reply]
The router step decides which tool/path to take. The tool describes itself as doing a web search for current information. 

I haven't done this myself yet, but I hope to soon-ish. I believe the step of selecting and using the tool is easy. I personally don't think the web results need to be vectorized, just scraped and returned to the LLM for analysis.

**Reply by cpursley** [reply]
Regarding the web results, I'd think that typical web pages (the useful parts of the text) would oftentimes end up being too many tokens if dumped right into the LLM context?

**Reply by brainlid** [reply]
You may be right. I was finally able to watch the YouTube video "[Reliable, fully local RAG agents with LLaMA3](https://www.youtube.com/watch?v=-ROS6gfYIts)" that you mentioned. It's cool. And yes, in that example web searches were vectorized and searched that way.

**Reply by cpursley** [reply]
> I personally don't think the web results need to be vectorized, just scraped and returned to the LLM for analysis.

I ended up have pretty good results with this approach using a larger context window model (Phi 3). Here's a little gist: https://gist.github.com/cpursley/b4af2ff3b56c912f659bd5300e422790#file-web_search-ex - the interesting part is probably the prompts.

What are you thoughts on creating some kind of WebSearch or WebSummarizer Chain?

**Reply by brainlid** [reply]
@cpursley, from what I've read or seen, I imagine it as a tool that's available to be called within a chain. And yes, I would like to have it. 🙂

**Reply by cpursley** [reply]
Cool. I need to refine it a good bit and make the search engine swappable as well as llm. Also, maybe have a text chunk vector retrieval option (instead of using llms to parse and summarize).
